Use ANSI sequences in colored output on Windows

GoaLitiuM opened the pull request:
Split Console to Windows specific and ANSI capable implementations, and detect the supported type at runtime. When no terminal is detected with `-color=on` option, ANSI colors will be used in order to prevent color information getting lost when output is redirected to other process.

Replaces #8618

GoaLitiuM committed:
Use ANSI sequences in colored output on Windows

Split Console to Windows specific and ANSI capable implementations, and detect the supported type at runtime. When no terminal is detected with -color=on option, ANSI colors will be used in order to prevent color information getting lost when output is redirected to other process.

CyberShadow reviewed (dismissed):
I think I mentioned this in the previous iteration, but currently, this looks like it confuses many things to achieve a very vaguely-defined goal.

- I can't think of any situation where it would make sense to use `ENABLE_VIRTUAL_TERMINAL_PROCESSING`. This flag looks useful for porting applications that already output ANSI control codes, and just need this one switch to enable Windows to parse them. But, DMD already is capable of using the native Windows console color APIs. Furthermore, if we are in a situation where we can set that flag, we could just use the native API instead.

- The original goal of the PR is to emit ANSI codes when error messages are going into a file or pipe, so build tools like Ninja or various CI services can print them out and humans could see colors in error messages. `ENABLE_VIRTUAL_TERMINAL_PROCESSING` is completely orthogonal to that, as far as DMD is concerned, because it is writing to a file or pipe, not a terminal.

- Another matter is allowing the user to choose which API is used for coloring messages. If there is a reason to ever default to ANSI output on Windows, then this setting would be orthogonal to the existing `-color` switch, but seeing as there isn't, a `-color=ansi` (with the implicit additional meaning to always use color in messages, whether the output is going to a file/pipe or terminal) would probably be sufficient.

- At a glance, the diff still looks bigger than what my gut feeling says it should be. Additionally, the suspiciously high number of `version (Windows)` and `version (Posix)` blocks indicate that things are not encapsulated as well as they could be, which gives me a feeling that perhaps there is room for improvement in the quality of implementation.

CyberShadow commented:
So, you're suggesting that `-color=on` should mean to emit ANSI codes on Windows if errors are not being written to a terminal. Although this would be inconsistent internally (whether colors are on or off is conceptually orthogonal to which API is used), I can see the benefit: `-color=on` would have no effect whatsoever with the Windows API when errors are redirected; plus, `-color=on` meaning to always emit ANSI codes would be consistent across all platforms, which would simplify writing build scripts.

CyberShadow commented:
Thinking about this some more, there is still the question of whether the ANSI codes will be displayable on the given system. After all, there are still two questions that affect this: whether the used Windows version supports parsing ANSI codes in terminal output, and whether this feature is enabled at the time that DMD's output is displayed (after being collected to a file/pipe).

That `-color=on` always enables color output does make sense from the perspective of a build script, to mean "in this configuration, DMD's messages will be collected by a tool that typically already handles tools which emit ANSI sequences", so I think we need a different way for users to say "my system supports ANSI displaying sequences and I can configure my terminal to do so at the moment when they are displayed".

Maybe an environment variable (something like `DMD_ANSI_COLOR_IN_MESSAGES`)?

GoaLitiuM commented:
If the user has explicitly requested `-color=on`, we should respect their choice and emit the color codes in a way that is appropriate to given context. So when the terminal is not detected, we should assume user (and their tooling) knows what they are doing and emit the color codes with ANSI sequences, which are not lost during redirection. 

[Ninja](https://github.com/ninja-build/ninja/blob/bf7107bb864d0383028202e3f4a4228c02302961/src/build.cc#L159) currently does strip away the ANSI sequences if the platform does not support displaying them, or if the user has not forced to emit these codes via environment variables. [Rust compiler](https://github.com/rust-lang/rust/commit/d293d1eea0b41e4ff65e8027e74eca9efd2dc7df) also seems to emit ANSI sequences on Windows with `-color always` option while terminal is not detected.

GoaLitiuM commented:
> 1. `-color=on` is not necessarily specified by the user. It can be part of the build script, which might have been tested only on a UNIX OS, and/or a Windows version that supports ANSI codes.

Build scripts can be fixed far more easily than restoring color information out of nothing.
> 2. We don't know that the "way that is appropriate to given context" means that we should be emitting ANSI codes, because we don't know how the output will be displayed. It might very well be a text file editor/viewer, or a web browser.

But we do know how the output will be displayed. The terminal detection method for Windows checks if the screen buffer is present, which is the only backwards compatible way to display color in Windows terminals. ANSI codes are widely accepted and recognized way to handle colors in these cases, so we should definitely use them when there is no other standardized ways to deal with them on Windows.

CyberShadow commented:
> Build scripts can be fixed far more easily than restoring color information out of nothing.

Why would you ever want to do that? Colors are a nice extra, they are never essential.

> ANSI codes are widely accepted and recognized way to handle colors in these cases, so we should definitely use them when there is no other standardized ways to deal with them on Windows.

I don't think one follows from the other.

It's much better to risk displaying messages without color, than to risk making them look like unreadable garbage due to unparsed color codes.

ZombineDev commented:
When a boolean flag is set to `on` the only sensible thing is to respect the decision. We should be in no business of patching up purely theoretical mistakes.
The only arguable question is whether we should have a separate Windows-only flag for forcing the coloring method. I don't see value in that (I think we should use ANSI, unless we're on interactive console on older versions of Windows), but I'm not opposed to it.

CyberShadow commented:
> When a boolean flag is set to `on` the only sensible thing is to respect the decision. We should be in no business of patching up purely theoretical mistakes.

What is this a reply to?

CyberShadow commented:
> I don't see value in that (I think we should use ANSI, unless we're on interactive console on older versions of Windows),

As mentioned above: the console that DMD runs on, and the console that its output will be displayed on, are not necessarily the same when DMD's messages are being redirected. Users of older Windows versions might have tools at their disposal for displaying those color codes regardless of what the Windows console supports.

CyberShadow commented:
As far as I can see, any way you turn it, there are going to be situations where the user will not get what they expect, whether it's garbage (when they expected readable error messages), or lack of color (which I think is much more acceptable). The only way to avoid this is to make it opt-in, because we just can't control in which circumstance DMD's messages are going to be displayed again. `-color=on` is not necessarily the best way to signal that opt-in, because otherwise there's no way to say "always use this platform's native color output" and variations thereof. An additional flag is not the best way either, because the compiler invocation can be in a build description or script under version control and thus difficult to modify locally, but this does not apply to an environment variable, which is what I suggested above.

CyberShadow commented:
> Users of older Windows versions might have tools at their disposal for displaying those color codes regardless of what the Windows console supports.

Just remembered one such tool - ConEmu optionally supports parsing ANSI sequences on all Windows versions it runs on.

GoaLitiuM commented:
> That's why I think we should use only ANSI for `--color=on` and only consider the Win API for `--color=auto`.

If you mean for terminal output, that's how I implemented it before but it was completely unnecessary as the old API still works in Windows 10, which is why I removed it. For redirected output, this is what the patch already does, and should be the expected behavior too since there is no other standardized way to represent colors in redirected console output.

@CyberShadow What you are proposing for this is adding another flag to control if we should use ANSI colors or not, but I see the current behavior of `-color=on` without terminal present a bug that only manifests on Windows: If I override the `auto` setting, I expect to see colors. You really want to fix a bug by adding an extra flag to enable so the option works as expected? Yes, this change might be breaking where the ANSI codes would unexpectedly show up because the user forgot they had colors enabled all the time. This would be more problematic before the new `-color=auto` default option was introduced, but the new `on` option on Windows does not do what it was advertised for, or put it simply, a bug.

CyberShadow commented:
> What you are proposing for this is adding another flag to control if we should use ANSI colors or not

That is not what I proposed. I proposed an environment variable.

> but I see the current behavior of `-color=on` without terminal present a bug that only manifests on Windows: If I override the `auto` setting, I expect to see colors.

We are going in circles.

You are not going to see colors when redirecting to a pipe, with this pull request or not, with `-color=on` or not, unless you are using a terminal / file viewer capable of parsing ANSI codes at the point when the redirected messages are displayed. We don't know when that is or what circumstances that will be in.

It is disingenuous to call this situation a bug, because it replaces one issue with another. Please review the discussion above.

I am interested in finding the correct approach to this situation which does not result in compromises resulting in a degraded experience for foreseeable situations. As far as I can see, using an environment variable to opt-in to ANSI output is the best approach in this discussion so far. If you have a better idea, or if you think another idea presented above is better, please present arguments in its case, but please do so taking into account the arguments presented above.

CyberShadow commented:
BTW, it would be nice to have colors work in Windows terminal emulators, like mintty. I had a quick look, and it looks like `isatty` from the DigitalMars C runtime returns `false` there, so it looks like we'd need to use a different detection method if we want to support them.

If we go with the environment variable route, it should also affect output when DMD thinks it's writing to a terminal, to support terminal emulators like MinTTY correctly.

CyberShadow commented:
I don't understand the point that I was trying to make three years ago.

In case of `-color=on`, we should enable the color method that is most appropriate for whatever is attached to `stdout`/`stderr`. If it's a Windows console, use the Windows console API; otherwise, we are forced to assume it's a pipe and use ANSI color codes.

WalterBright commented:
The implementation of isatty for DMC:

https://github.com/DigitalMars/dmc/blob/master/src/WIN32/ISATTY.C

Not much to it.

GoaLitiuM commented:
Closing this in favor of #12541
